<title>README</title>

<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type="text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<div class="container-fluid main-container">

<!-- tabsets -->
<script src="data:application/x-javascript;base64,Cgp3aW5kb3cuYnVpbGRUYWJzZXRzID0gZnVuY3Rpb24odG9jSUQpIHsKCiAgLy8gYnVpbGQgYSB0YWJzZXQgZnJvbSBhIHNlY3Rpb24gZGl2IHdpdGggdGhlIC50YWJzZXQgY2xhc3MKICBmdW5jdGlvbiBidWlsZFRhYnNldCh0YWJzZXQpIHsKCiAgICAvLyBjaGVjayBmb3IgZmFkZSBhbmQgcGlsbHMgb3B0aW9ucwogICAgdmFyIGZhZGUgPSB0YWJzZXQuaGFzQ2xhc3MoInRhYnNldC1mYWRlIik7CiAgICB2YXIgcGlsbHMgPSB0YWJzZXQuaGFzQ2xhc3MoInRhYnNldC1waWxscyIpOwogICAgdmFyIG5hdkNsYXNzID0gcGlsbHMgPyAibmF2LXBpbGxzIiA6ICJuYXYtdGFicyI7CgogICAgLy8gZGV0ZXJtaW5lIHRoZSBoZWFkaW5nIGxldmVsIG9mIHRoZSB0YWJzZXQgYW5kIHRhYnMKICAgIHZhciBtYXRjaCA9IHRhYnNldC5hdHRyKCdjbGFzcycpLm1hdGNoKC9sZXZlbChcZCkgLyk7CiAgICBpZiAobWF0Y2ggPT09IG51bGwpCiAgICAgIHJldHVybjsKICAgIHZhciB0YWJzZXRMZXZlbCA9IE51bWJlcihtYXRjaFsxXSk7CiAgICB2YXIgdGFiTGV2ZWwgPSB0YWJzZXRMZXZlbCArIDE7CgogICAgLy8gZmluZCBhbGwgc3ViaGVhZGluZ3MgaW1tZWRpYXRlbHkgYmVsb3cKICAgIHZhciB0YWJzID0gdGFic2V0LmZpbmQoImRpdi5zZWN0aW9uLmxldmVsIiArIHRhYkxldmVsKTsKICAgIGlmICghdGFicy5sZW5ndGgpCiAgICAgIHJldHVybjsKCiAgICAvLyBjcmVhdGUgdGFibGlzdCBhbmQgdGFiLWNvbnRlbnQgZWxlbWVudHMKICAgIHZhciB0YWJMaXN0ID0gJCgnPHVsIGNsYXNzPSJuYXYgJyArIG5hdkNsYXNzICsgJyIgcm9sZT0idGFibGlzdCI+PC91bD4nKTsKICAgICQodGFic1swXSkuYmVmb3JlKHRhYkxpc3QpOwogICAgdmFyIHRhYkNvbnRlbnQgPSAkKCc8ZGl2IGNsYXNzPSJ0YWItY29udGVudCI+PC9kaXY+Jyk7CiAgICAkKHRhYnNbMF0pLmJlZm9yZSh0YWJDb250ZW50KTsKCiAgICAvLyBidWlsZCB0aGUgdGFic2V0CiAgICB0YWJzLmVhY2goZnVuY3Rpb24oaSkgewoKICAgICAgLy8gZ2V0IHRoZSB0YWIgZGl2CiAgICAgIHZhciB0YWIgPSAkKHRhYnNbaV0pOwoKICAgICAgLy8gZ2V0IHRoZSBpZCB0aGVuIHNhbml0aXplIGl0IGZvciB1c2Ugd2l0aCBib290c3RyYXAgdGFicwogICAgICB2YXIgaWQgPSB0YWIuYXR0cignaWQnKTsKCiAgICAgIC8vIHJlbW92ZSBhbnkgdGFibGUgb2YgY29udGVudHMgZW50cmllcyBhc3NvY2lhdGVkIHdpdGgKICAgICAgLy8gdGhpcyBJRCAoc2luY2Ugd2UnbGwgYmUgcmVtb3ZpbmcgdGhlIGhlYWRpbmcgZWxlbWVudCkKICAgICAgJCgiZGl2IyIgKyB0b2NJRCArICIgbGkgYVtocmVmPScjIiArIGlkICsgIiddIikucGFyZW50KCkucmVtb3ZlKCk7CgogICAgICAvLyBzYW5pdGl6ZSB0aGUgaWQgZm9yIHVzZSB3aXRoIGJvb3RzdHJhcCB0YWJzCiAgICAgIGlkID0gaWQucmVwbGFjZSgvWy5cLz8mISM8Pl0vZywgJycpLnJlcGxhY2UoL1xzL2csICdfJyk7CiAgICAgIHRhYi5hdHRyKCdpZCcsIGlkKTsKCiAgICAgIC8vIGdldCB0aGUgaGVhZGluZyBlbGVtZW50IHdpdGhpbiBpdCwgZ3JhYiBpdCdzIHRleHQsIHRoZW4gcmVtb3ZlIGl0CiAgICAgIHZhciBoZWFkaW5nID0gdGFiLmZpbmQoJ2gnICsgdGFiTGV2ZWwgKyAnOmZpcnN0Jyk7CiAgICAgIHZhciBoZWFkaW5nVGV4dCA9IGhlYWRpbmcuaHRtbCgpOwogICAgICBoZWFkaW5nLnJlbW92ZSgpOwoKICAgICAgLy8gYnVpbGQgYW5kIGFwcGVuZCB0aGUgdGFiIGxpc3QgaXRlbQogICAgICB2YXIgYSA9ICQoJzxhIHJvbGU9InRhYiIgZGF0YS10b2dnbGU9InRhYiI+JyArIGhlYWRpbmdUZXh0ICsgJzwvYT4nKTsKICAgICAgYS5hdHRyKCdocmVmJywgJyMnICsgaWQpOwogICAgICBhLmF0dHIoJ2FyaWEtY29udHJvbHMnLCBpZCk7CiAgICAgIHZhciBsaSA9ICQoJzxsaSByb2xlPSJwcmVzZW50YXRpb24iPjwvbGk+Jyk7CiAgICAgIGxpLmFwcGVuZChhKTsKICAgICAgaWYgKGkgPT09IDApCiAgICAgICAgbGkuYXR0cignY2xhc3MnLCAnYWN0aXZlJyk7CiAgICAgIHRhYkxpc3QuYXBwZW5kKGxpKTsKCiAgICAgIC8vIHNldCBpdCdzIGF0dHJpYnV0ZXMKICAgICAgdGFiLmF0dHIoJ3JvbGUnLCAndGFicGFuZWwnKTsKICAgICAgdGFiLmFkZENsYXNzKCd0YWItcGFuZScpOwogICAgICB0YWIuYWRkQ2xhc3MoJ3RhYmJlZC1wYW5lJyk7CiAgICAgIGlmIChmYWRlKQogICAgICAgIHRhYi5hZGRDbGFzcygnZmFkZScpOwogICAgICBpZiAoaSA9PT0gMCkgewogICAgICAgIHRhYi5hZGRDbGFzcygnYWN0aXZlJyk7CiAgICAgICAgaWYgKGZhZGUpCiAgICAgICAgICB0YWIuYWRkQ2xhc3MoJ2luJyk7CiAgICAgIH0KCiAgICAgIC8vIG1vdmUgaXQgaW50byB0aGUgdGFiIGNvbnRlbnQgZGl2CiAgICAgIHRhYi5kZXRhY2goKS5hcHBlbmRUbyh0YWJDb250ZW50KTsKICAgIH0pOwogIH0KCiAgLy8gY29udmVydCBzZWN0aW9uIGRpdnMgd2l0aCB0aGUgLnRhYnNldCBjbGFzcyB0byB0YWJzZXRzCiAgdmFyIHRhYnNldHMgPSAkKCJkaXYuc2VjdGlvbi50YWJzZXQiKTsKICB0YWJzZXRzLmVhY2goZnVuY3Rpb24oaSkgewogICAgYnVpbGRUYWJzZXQoJCh0YWJzZXRzW2ldKSk7CiAgfSk7Cn07Cgo="></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="fluid-row" id="header">



<h1 class="title toc-ignore">README</h1>

</div>


<div id="installation" class="section level2">
<h2>Installation</h2>
<p>We have tested on GCC-4.8, 4.9 and 5.1.</p>
<pre class="r"><code>### if on Broad Institute servers, add GCC-5.1
reuse -q GCC-5.1

############## DOWNLOAD BOOST (if not installed) ###############
wget https://sourceforge.net/projects/boost/files/boost/1.61.0/boost_1_61_0.tar.gz
tar -xvzf boost_1_61_0.tar.gz
## we only user header-only libraries, so no compiling of Boost is needed

############### DOWNLOAD VARIANT BAM ############### 
git clone --recursive https://github.com/broadinstitute/SnowmanSVgit
cd SnowmanSV

############### COMPILE AND INSTALL ###############
./configure --with-boost=&lt;path_to_boost&gt;  ## e.g. ~/boost_1_61_0
make

############### QUICK START ############### 
# run tumor / normal on Chr22, with 4 cores
SnowmanSV/src/Snowman/snowman -t tumor.bam -n normal.bam -k 22 -G ref.fa -a test_id -p -4

## get help
snowman --help</code></pre>
</div>
<div id="description" class="section level2">
<h2>Description</h2>
<p>Snowman is a method for detecting structural variants in sequencing data using genome-wide local assembly. Under the hood, Snowman uses a custom implementation of <a href="https://github.com/jts/sga">SGA</a> (String Graph Assembler) by Jared Simpson, and <a href="https://github.com/lh3/bwa">BWA-MEM</a> by Heng Li. Contigs are assembled for every 25 kb window (with some small overlap) for every region in the genome. The default is to use only clipped, discordant, unmapped and indel reads, although this can be customized to any set of reads at the command line using <a href="https://github.com/jwalabroad/VariantBam">VariantBam</a> rules. These contigs are then immediately aligned to the reference with BWA-MEM and parsed to identify variants. Sequencing reads are likewise realigned to the contigs with BWA-MEM, and variants are scored.</p>
</div>
<div id="scope" class="section level2">
<h2>Scope</h2>
<p>Snowman is currently configured to provide indel and rearrangement calls (and anything “in between”). It has been most widely tested as a somatic variant caller, and outputs separate VCFs for somatic and germline. If only a single BAM is present, input with the -t flag. In this case, the results will contain all calls, with no germline/somatic designation.</p>
</div>
<div id="required-inputs" class="section level2">
<h2>Required Inputs</h2>
<p>Any number of BAM/SAM/CRAM files can be supplied at once. Snowman uses random access of the BAMs to obtain pair-mate reads, and so requires the files to be indexed (samtools index). Tumor BAMs are input with -t and normal with -n. The order does not matter. At least one “tumor” BAM is required, although this could be just a single germline sample, or paired with a set of parents input with -n to search for de novo alterations. All assemblies are done jointly, combining reads across all of the BAM files. The source of the variant support reads is then tracked during realignment of reads to assembly contigs. A BWA indexed reference genome must be supplied as well (-G).</p>
</div>
<div id="output-file-description" class="section level2">
<h2>Output file description</h2>
<ul>
<li><p><code>*.bps.txt.gz</code> Raw, unfiltered variants. This file is parsed at the end to produce the VCF files. With the bps.txt.gz, one can define a new set of filteirng criteria (depending on sensitivity/specificity needs) using <code>snowman refilter</code>.</p></li>
<li><p><code>*.contigs.bam</code> All assembly contigs as aligned to the reference with BWA-MEM. Note that this is an unsorted file. To view in IGV, it must be first sorted and indexed (e.g. <code>samtools sort -m 8G id.contigs.bam id.sort &amp;&amp; samtools index id.sort.bam</code>)</p></li>
<li><p><code>*.discordants.txt.gz</code> Information on all clusters of discordant reads identified with 2+ reads.</p></li>
<li><p><code>*.log</code> Log file giving run-time information, including CPU and Wall time (and how it was partitioned among the tasks), number of reads retrieved and contigs assembled for each region.</p></li>
<li><p><code>*.alignments.txt.gz</code> An ASCII plot of variant-supporting contigs and the BWA-MEM alignment of reads to the contigs. This file is incredibly useful for debugging and visually inspecting the exact information Snowman saw when it performed the variant-calling. This file is typically quite large. The recommended usage is to identify the contig name of your variant of interest first from the VCF file (SCTG=contig_name). Then do <code>gunzip -c id.alignment.txt.gz | grep contig_name &gt; plot.txt</code>. It is highly recommended that you view in a text editor with line truncation turned OFF, so as to not jumble the alignments.</p></li>
<li><p><code>*.cigarmap.txt.gz</code> Information on clusters of indel alignments as identified from the original BAMs.</p></li>
<li><p><code>*.vcf</code> VCF of rearrangements and indels parsed from bps.txt.gz and with a somatic_score == 1 (somatic) or 0 (germline) and quality == PASS. <em>NOTE</em> that the cutoff for rearrangement vs indel is taken from BWA-MEM, whether it produces a single gapped-alignment or two separate alignments. This is an arbitrary cutoff, just as there is no clear consensus distinction between what constitutes an “indel” and a “structural variant”. The unfiltered VCF files include non-pass variants.</p></li>
</ul>
</div>
<div id="filtering-refiltering" class="section level2">
<h2>Filtering / Refiltering</h2>
<p>Snowman performs a series of log-likelihood calculations for each variant. The purpose is to first classify a variant as real vs artifact, and then to determine if the variant is somatic or germline. These log-likelihoods are output in the VCF and bps.txt.gz file and described here: * <code>LOD (LO)</code> - Log of the odds that variant is real vs artifact. For indels, the likelihood of an artifact read is proportional to the length of local repeats (repeating units up to 5 long per unit) * <code>LR</code> - Log of the odds that the variant has allelic fraction (AF) of 0 or &gt;=0.5. This is used for somatic vs germline classification * <code>SL</code> - Scaled LOD. LOD scores is heuristically scaled as: (min(Mapping quality #1, Mapping quality #2) - 2 * NM) / 60 * LOD</p>
<p>Snowman can refilter the bps.txt.gz file to produce new VCFs with different stringency cutoffs. To run, the following are required: * <code>-b</code> - a BAM from the original run, which is used just for its header * <code>-i</code> - input bps.txt.gz file</p>
</div>
<div id="auxillary-tools" class="section level2">
<h2>Auxillary Tools</h2>
<ul>
<li><code>snowman benchmark</code></li>
</ul>
<p>Snowman ships with tools for testing the assemblies and for generating in-silico tumor genomes for testing.</p>
<ul>
<li><code>--sim-breaks-power</code> Simulates contigs containing structural variants.</li>
</ul>
<pre><code>### Simulate SVs and indels from the reference.
-R &lt;num_rearrangements&gt; -X &lt;num_indels&gt; --add-scrambled-inserts
</code></pre>
<ul>
<li><code>--split-bam</code> Simple tool for splitting a BAM file randomly into smaller fragments. This is useful for generating a sub-sampled BAM, and ensuring that each of the subsampled BAMs have different reads in them. <code>split-bam</code> will use the read name (and a seed provied with <code>-s</code>) to generate the random numbers, which ensures that read-pairs are always sent to the same BAM.</li>
</ul>
<pre><code>### split BAM into 25% and 50% pieces at certain regions
snowman benchmark --split-bam -b &lt;in.bam&gt; -f 0.2,0.5 -k &lt;regions.bed&gt;
### split the entire BAM
snowman benchmark --split-bam -b &lt;in.bam&gt; -f 0.2,0.5 </code></pre>
<ul>
<li><p><code>--realign-test</code> Test the abiltiy of BWA-MEM to realign contigs of different lengths, with errors. Useful for gauging how often true indels will be missed during the BWA-MEM realignment phase.</p>
<pre><code>snowman benchmark --realign-test</code></pre></li>
<li><p><code>--realign-sv-test</code> Test the abiltiy of BWA-MEM to realign SV contigs of different lengths, with errors. Useful for gauging how often true rearrangements will be missed during the BWA-MEM realignment phase.</p>
<pre><code>snowman benchmark --realign-sv-test</code></pre></li>
</ul>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>