---
title: "RECIPES"
output: 
  html_document:
  toc: true
  toc_depth: 2
---

```{r set-options, echo=FALSE, cache=FALSE}
options(width = 10000)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Running Snowman

#### Whole-genome Somatic SV and Indel Detection with 4 cores
```{r, eval=FALSE}
wget "https://data.broadinstitute.org/snowman/dbsnp_indel.vcf"
DBSNP=dbsnp_indel.vcf
G=/seq/references/Homo_sapiens_assembly19/v1/Homo_sapiens_assembly19.fasta
## -a is any string you like, which gives the run an ID
snowman run -t $TUM_BAM -n $NORM_BAM -p $CORES -D $DBSNP -a somatic_run -G $REF
```

#### Whole-genome Germline SV and Indel Detection
```{r, eval=FALSE}
## Set -I to not do mate-region lookup if mates are mapped to different chromosome.
##   This is appropriate for germline-analysis, where we don't have a built-in control
##   and this will bog down the mate-lookup in spurious mapping artifacts
## 
## Set -L to 6 which means that 6 or more mate reads must be clustered to 
##   trigger a mate lookup. This also reduces spurious lookups, and is more 
##   appropriate the expected ALT counts found in a germline sample 
##   (as opposed to impure, subclonal events in cancer).
snowman run -t $GERMLINE_BAM -p $CORES -L 6 -I -a germline_run -G $REF
```

#### Run a local assembly
```{r, eval=FALSE}
## -k can be a chromosome, a samtools/IGV style string 
##     (e.g. 1:1,000,000-2,000,000), or a BED file
k=chr17:7,541,145-7,621,399
snowman run -t $TUM_BAM -n $NORM_BAM -p $CORES -k $k  -a TP53 -G $REF
```

#### Assemble in chunks of 100kb
```{r, eval=FALSE}
snowman run -t $TUM_BAM -n $NORM_BAM -p $CORES -c 100000 -G $REF -a somatic_100kb
```

#### Assemble ALL reads (not just clipped/discordant/gapped/unmapped reads)
```{r, eval=FALSE}
snowman run -t $TUM_BAM -n $NORM_BAM -p $CORES -r all -G $REF -a somatic_all_reads
```

#### Decrease error-tolerance for high coverage of high quality BAM
```{r, eval=FALSE}
## default error tolerance is 0.025 (see SGA -e parameter), tolerating 2.5% divergence
snowman run -t $TUM_BAM -n $NORM_BAM -e 0.015 -a somatic_run_e15 -p $CORES -D $DBSNP -G $REF
```

#### Get a real-time snapshot of where Snowman is while running
```{r, eval=FALSE}
tail somatic_run.log
```

## Analyzing output

#### Debug a local assembly and produce the assembly graph
```{r, eval=FALSE}
k=chr17:7,541,145-7,621,399
snowman run -t $BAM -a local_test -k $k --write-asqg

## plot the graph
$GIT/SnowmanSV/R/snow-asqg.R
```

#### View the filtered results
```{r, eval=FALSE}
less somatic_run.snowman.somatic.sv.vcf
less somatic_run.snowman.somatic.indel.vcf
```

#### View all of the ASCII alignments (warning, can be large)
```{r, eval=FALSE}
## Make a read-only and no-line-wrapping version of emacs. 
## Very useful for *.alignments.txt.gz files
function ev { 
  emacs $1 --eval '(setq buffer-read-only t)' -nw --eval '(setq truncate-lines t)';
  }
ev somatic_run.alignments.txt.gz 
```

#### View a particular contig with read-to-contig alignments
```{r, eval=FALSE}
gunzip -c somatic_run.alignments.txt.gz | grep c_1_123456789_123476789 > c_1_123456789_123476789.alignments.txt
ev c_1_123456789_123476789.alignments.txt
```

#### Make a function to sort and index the contigs.bam file (to enable IGV viewing)
```{r, eval=FALSE}
function sai() {
  if [[ -f $1.contigs.bam ]]; then
     samtools sort -m 4G $1.contigs.bam -o $1.contigs.sort.bam
     mv $1.contigs.sort.bam $1.contigs.bam
     samtools index $1.contigs.bam
  fi
}

sai somatic_run
```