\section{Aligned\-Contig Class Reference}
\label{classAlignedContig}\index{AlignedContig@{AlignedContig}}
{\tt \#include $<$Aligned\-Contig.h$>$}

\subsection*{Public Member Functions}
\begin{CompactItemize}
\item 
\bf{Aligned\-Contig} (const std::string \&sam, const Bam\-Tools::Bam\-Reader $\ast$reader, const Snow\-Tools::Genomic\-Region \&twindow, const Cigar\-Map \&nmap, const Cigar\-Map \&tmap)
\item 
\textbf{Aligned\-Contig} (const Bam\-Tools::Bam\-Alignment \&a)\label{classAlignedContig_4b4f577516be62f5f0799f66427cdf84}

\item 
bool \bf{is\-Worse} (const \bf{Aligned\-Contig} \&ac) const 
\item 
void \bf{add\-Alignment} (const Bam\-Tools::Bam\-Alignment \&align, const Snow\-Tools::Genomic\-Region \&\bf{window}, const Cigar\-Map \&nmap, const Cigar\-Map \&tmap)
\item 
void \textbf{add\-Alignment} (const Bam\-Tools::Bam\-Alignment \&align)\label{classAlignedContig_bedec8ec48a3179f35d813fb7c013694}

\item 
void \bf{add\-Discordant\-Cluster} (Discordant\-Cluster dc)\label{classAlignedContig_88a303a9bc8cc07a7efa4548cb73e5fd}

\begin{CompactList}\small\item\em add a discordant cluster that maps to same regions as this contig \item\end{CompactList}\item 
std::string \bf{print\-Discordant\-Clusters} () const 
\item 
std::string \bf{get\-Contig\-Name} () const \label{classAlignedContig_87074cde30e8ecc61ab848969a7ecc9d}

\begin{CompactList}\small\item\em return the name of the contig \item\end{CompactList}\item 
int \bf{get\-Max\-Mapq} () const 
\item 
int \bf{get\-Min\-Mapq} () const 
\item 
void \bf{split\-Coverage} ()
\item 
void \bf{blacklist} (GRC \&grv)
\item 
bool \bf{parse\-Discovar\-Name} (size\_\-t \&tumor, size\_\-t \&normal)
\item 
void \bf{print\-Contig\-Fasta} (std::ofstream \&os) const 
\item 
void \bf{set\-Multi\-Map\-Break\-Pairs} ()
\item 
void \bf{align\-Reads\-To\-Contigs} (Read\-Vec \&bav)
\item 
std::string \bf{get\-Sequence} () const \label{classAlignedContig_54f942ec1c6d4d9f5c5473d145f5fb8a}

\begin{CompactList}\small\item\em return the contig sequence as it came off the assembler \item\end{CompactList}\item 
bool \bf{has\-Sub\-Sequence} (const std::string \&subseq) const \label{classAlignedContig_f440054db1ee3fabbf7c21a7f6ae5d20}

\begin{CompactList}\small\item\em detemine if the contig contains a subsequence \item\end{CompactList}\item 
bool \bf{has\-Variant} () const 
\item 
std::vector$<$ Break\-Point $>$ \bf{get\-All\-Break\-Points} () const 
\item 
std::vector$<$ const Break\-Point $\ast$ $>$ \textbf{get\-All\-Break\-Point\-Pointers} () const \label{classAlignedContig_eb8f9a66cb13b2e2faa56bacf63cf4ad}

\end{CompactItemize}
\subsection*{Public Attributes}
\begin{CompactItemize}
\item 
std::string \bf{samrecord}
\item 
Snow\-Tools::Genomic\-Region \bf{window}
\item 
std::vector$<$ Break\-Point $>$ \textbf{m\_\-local\_\-breaks}\label{classAlignedContig_bf3da53e79235032f52481056f2ad1e1}

\item 
Break\-Point \textbf{m\_\-global\_\-bp}\label{classAlignedContig_aaf8303226b21cd895ff11d55ab0914c}

\item 
Read\-Vec \textbf{m\_\-bamreads}\label{classAlignedContig_af838fff09dd3164c41035afd6b18c27}

\item 
bool \textbf{m\_\-skip} = false\label{classAlignedContig_6485296807f802f3fe2b049d24b47dac}

\item 
Alignment\-Fragment\-Vector \textbf{m\_\-align}\label{classAlignedContig_6372374f9b177011a2c6b43bab9cd4f7}

\end{CompactItemize}
\subsection*{Friends}
\begin{CompactItemize}
\item 
std::ostream \& \bf{operator$<$$<$} (std::ostream \&out, const \bf{Aligned\-Contig} \&ac)\label{classAlignedContig_064b72b01461a676f213a56f80f67d87}

\begin{CompactList}\small\item\em print this contig \item\end{CompactList}\end{CompactItemize}


\subsection{Detailed Description}
Contains the mapping of an aligned contig to the reference genome, along with pointer to all of the reads aligned to this contig, and a store of all of the breakpoints associated with this contig 



\subsection{Constructor \& Destructor Documentation}
\index{AlignedContig@{Aligned\-Contig}!AlignedContig@{AlignedContig}}
\index{AlignedContig@{AlignedContig}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}Aligned\-Contig::Aligned\-Contig (const std::string \& {\em sam}, const Bam\-Tools::Bam\-Reader $\ast$ {\em reader}, const Snow\-Tools::Genomic\-Region \& {\em twindow}, const Cigar\-Map \& {\em nmap}, const Cigar\-Map \& {\em tmap})}\label{classAlignedContig_bee5720be1588fc0f5e2033c68e47f4b}


Constructor which parses an alignment record from BWA (a potentially multi-line SAM record) \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em const}]reference to a string representing a SAM alignment (contains newlines if multi-part alignment) \item[{\em const}]pointer to a Bam\-Reader, which is used to convert chr ids to strings (e.g. X, Y) \item[{\em const}]reference to a Snow\-Tools::Genomic\-Region window specifying where in the reference this contig was assembled from. \end{description}
\end{Desc}


\subsection{Member Function Documentation}
\index{AlignedContig@{Aligned\-Contig}!addAlignment@{addAlignment}}
\index{addAlignment@{addAlignment}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Aligned\-Contig::add\-Alignment (const Bam\-Tools::Bam\-Alignment \& {\em align}, const Snow\-Tools::Genomic\-Region \& {\em window}, const Cigar\-Map \& {\em nmap}, const Cigar\-Map \& {\em tmap})}\label{classAlignedContig_adc92b588d491d20316f23d3adebe27c}


Get whether the query is on the reverse strand \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em b}]pointer to an alignment \end{description}
\end{Desc}
\begin{Desc}
\item[Returns:]boolean true if query is on the reverse strand \end{Desc}
\index{AlignedContig@{Aligned\-Contig}!alignReadsToContigs@{alignReadsToContigs}}
\index{alignReadsToContigs@{alignReadsToContigs}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Aligned\-Contig::align\-Reads\-To\-Contigs (Read\-Vec \& {\em bav})}\label{classAlignedContig_0ad74ad41bc628ebb08ba95c1b6e6376}


align reads to contig and modify their tags to show contig mapping Currently this function will attempt a Smith\-Waterman alignment for all reads that don't have an exact mapping to the contig. \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em bav}]Vector of read smart pointers to align. Modifies their SW tag \end{description}
\end{Desc}
\index{AlignedContig@{Aligned\-Contig}!blacklist@{blacklist}}
\index{blacklist@{blacklist}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Aligned\-Contig::blacklist (GRC \& {\em grv})}\label{classAlignedContig_b3889f82451c51a0a8c2946019af6c1a}


Checks if any of the indel breaks are in a blacklist. If so, mark the breakpoints of the indels for skipping. That is, has\-Minmal() will return false; \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em grv}]The blaclist regions \end{description}
\end{Desc}
\index{AlignedContig@{Aligned\-Contig}!getAllBreakPoints@{getAllBreakPoints}}
\index{getAllBreakPoints@{getAllBreakPoints}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}std::vector$<$Break\-Point$>$ Aligned\-Contig::get\-All\-Break\-Points () const}\label{classAlignedContig_0e801da1a6657287362fc291c72a40f3}


retrieves all of the breakpoints by combining indels with global mutli-map break \begin{Desc}
\item[Returns:]vector of ind \end{Desc}
\index{AlignedContig@{Aligned\-Contig}!getMaxMapq@{getMaxMapq}}
\index{getMaxMapq@{getMaxMapq}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}int Aligned\-Contig::get\-Max\-Mapq () const\hspace{0.3cm}{\tt  [inline]}}\label{classAlignedContig_6e859c97339b1211923816d20d76380c}


get the maximum mapping quality from all alignments \begin{Desc}
\item[Returns:]int max mapq \end{Desc}
\index{AlignedContig@{Aligned\-Contig}!getMinMapq@{getMinMapq}}
\index{getMinMapq@{getMinMapq}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}int Aligned\-Contig::get\-Min\-Mapq () const\hspace{0.3cm}{\tt  [inline]}}\label{classAlignedContig_d339b798afa99c0abcf1a2b5dd27642f}


get the minimum mapping quality from all alignments \begin{Desc}
\item[Returns:]int min mapq \end{Desc}
\index{AlignedContig@{Aligned\-Contig}!hasVariant@{hasVariant}}
\index{hasVariant@{hasVariant}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}bool Aligned\-Contig::has\-Variant () const}\label{classAlignedContig_840897416cdf4e5444b94b3ecfb303ca}


query if this contig contains a potential variant (indel or multi-map) \begin{Desc}
\item[Returns:]true if there is multimapping or an indel \end{Desc}
\index{AlignedContig@{Aligned\-Contig}!isWorse@{isWorse}}
\index{isWorse@{isWorse}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}bool Aligned\-Contig::is\-Worse (const \bf{Aligned\-Contig} \& {\em ac}) const}\label{classAlignedContig_9d4ebb5861cf988ffe1e9aa22715da7f}


Determine if this contig has identical breaks and is better than another. \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em const}]reference to another \doxyref{Aligned\-Contig}{p.}{classAlignedContig} \end{description}
\end{Desc}
\begin{Desc}
\item[Returns:]bool bool returning true iff this contig has identical info has better MAPQ, or equal MAPQ but longer \end{Desc}
\index{AlignedContig@{Aligned\-Contig}!parseDiscovarName@{parseDiscovarName}}
\index{parseDiscovarName@{parseDiscovarName}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}bool Aligned\-Contig::parse\-Discovar\-Name (size\_\-t \& {\em tumor}, size\_\-t \& {\em normal})}\label{classAlignedContig_af0765a3cf966158ce94a423682f4e4a}


if this is a Discovar contig, extract the tumor and normal read support \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em reference}]to int to fill for normal support \item[{\em reference}]to int to fill for tumor support \end{description}
\end{Desc}
\begin{Desc}
\item[Returns:]boolean reporting if this is a discovar name \end{Desc}
\index{AlignedContig@{Aligned\-Contig}!printContigFasta@{printContigFasta}}
\index{printContigFasta@{printContigFasta}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Aligned\-Contig::print\-Contig\-Fasta (std::ofstream \& {\em os}) const}\label{classAlignedContig_de3eff377e9359b6a0a82075530ca638}


dump the contigs to a fasta \begin{Desc}
\item[Parameters:]
\begin{description}
\item[{\em ostream}]to write to \end{description}
\end{Desc}
\index{AlignedContig@{Aligned\-Contig}!printDiscordantClusters@{printDiscordantClusters}}
\index{printDiscordantClusters@{printDiscordantClusters}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}std::string Aligned\-Contig::print\-Discordant\-Clusters () const}\label{classAlignedContig_d92b57eafadc8f619a2d9a68a3fee37b}


loop through the vector of Discordant\-Cluster objects associated with this contig and print \index{AlignedContig@{Aligned\-Contig}!setMultiMapBreakPairs@{setMultiMapBreakPairs}}
\index{setMultiMapBreakPairs@{setMultiMapBreakPairs}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Aligned\-Contig::set\-Multi\-Map\-Break\-Pairs ()}\label{classAlignedContig_d6d2f1dd0a149612a5c9d803a9b6ba85}


set the breakpoints on the reference by combining multi-mapped contigs \index{AlignedContig@{Aligned\-Contig}!splitCoverage@{splitCoverage}}
\index{splitCoverage@{splitCoverage}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}void Aligned\-Contig::split\-Coverage ()}\label{classAlignedContig_e5681bd6d2185c4f274d68f83404e072}


loop through all of the breakpoints and calculate the split read support for each. Requires aligned\-Reads to have been run first (will error if not run). 

\subsection{Member Data Documentation}
\index{AlignedContig@{Aligned\-Contig}!samrecord@{samrecord}}
\index{samrecord@{samrecord}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}std::string \bf{Aligned\-Contig::samrecord}}\label{classAlignedContig_a3b7e07742b8d626591bd2078be88920}


the original SAM record \index{AlignedContig@{Aligned\-Contig}!window@{window}}
\index{window@{window}!AlignedContig@{Aligned\-Contig}}
\subsubsection{\setlength{\rightskip}{0pt plus 5cm}Snow\-Tools::Genomic\-Region \bf{Aligned\-Contig::window}}\label{classAlignedContig_a94a9f382b760c02333dd2b3085c8c8d}


reference window from where this contig was assembled 

The documentation for this class was generated from the following file:\begin{CompactItemize}
\item 
/xchip/gistic/Jeremiah/GIT/Snowman\-SV/src/Snowman/Aligned\-Contig.h\end{CompactItemize}
