<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>SnowmanSV: /xchip/gistic/Jeremiah/GIT/SnowmanSV/src/Snowman/BamAndReads.h Source File</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.4.7 -->
<div class="tabs">
  <ul>
    <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
    <li><a href="annotated.html"><span>Classes</span></a></li>
    <li id="current"><a href="files.html"><span>Files</span></a></li>
  </ul></div>
<h1>/xchip/gistic/Jeremiah/GIT/SnowmanSV/src/Snowman/BamAndReads.h</h1><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 <span class="preprocessor">#ifndef BAM_AND_READS_H</span>
<a name="l00002"></a>00002 <span class="preprocessor"></span><span class="preprocessor">#define BAM_AND_READS_H</span>
<a name="l00003"></a>00003 <span class="preprocessor"></span>
<a name="l00004"></a>00004 <span class="preprocessor">#include &lt;vector&gt;</span>
<a name="l00005"></a>00005 <span class="preprocessor">#include "SnowTools/GenomicRegion.h"</span>
<a name="l00006"></a>00006 <span class="preprocessor">#include "SnowTools/GenomicRegionVector.h"</span>
<a name="l00007"></a>00007 <span class="preprocessor">#include "VariantBamReader.h"</span>
<a name="l00008"></a>00008 <span class="preprocessor">#include "MiniRules.h"</span>
<a name="l00009"></a>00009 <span class="preprocessor">#include &lt;time.h&gt;</span>
<a name="l00010"></a>00010 
<a name="l00011"></a>00011 <span class="preprocessor">#include "reads.h"</span>
<a name="l00012"></a>00012 
<a name="l00013"></a>00013 <span class="comment">//#define FORGORDON 1</span>
<a name="l00014"></a>00014 
<a name="l00015"></a>00015 <span class="preprocessor">#ifndef FORGORDON</span>
<a name="l00016"></a>00016 <span class="preprocessor"></span><span class="preprocessor">#include "Coverage.h"</span>
<a name="l00017"></a>00017 <span class="preprocessor">#endif</span>
<a name="l00018"></a>00018 <span class="preprocessor"></span>
<a name="l00019"></a>00019 <span class="keyword">using namespace </span>std;
<a name="l00020"></a>00020 <span class="keyword">typedef</span> unordered_map&lt;string, size_t&gt; CigarMap;
<a name="l00021"></a>00021 
<a name="l00022"></a>00022 <span class="comment">// a smaller region to assemble</span>
<a name="l00023"></a>00023 <span class="keyword">struct </span>AssemblyRegion {
<a name="l00024"></a>00024 
<a name="l00025"></a>00025   AssemblyRegion() {}
<a name="l00026"></a>00026   AssemblyRegion(GenomicRegion gr) { region = gr; }
<a name="l00027"></a>00027   ~AssemblyRegion() {
<a name="l00028"></a>00028   }
<a name="l00029"></a>00029   
<a name="l00030"></a>00030   ReadVec reads;
<a name="l00031"></a>00031   GenomicRegion region;
<a name="l00032"></a>00032 
<a name="l00033"></a>00033   <span class="comment">//GenomicRegionVector partner_windows;</span>
<a name="l00034"></a>00034   std::vector&lt;GenomicRegion&gt; partner_windows;
<a name="l00035"></a>00035   GenomicRegionVector partner_windows_full;
<a name="l00036"></a>00036   <span class="comment">//GenomicIntervalTreeMap tree_pw;</span>
<a name="l00037"></a>00037 
<a name="l00038"></a>00038   <span class="keywordtype">void</span> removeBlacklist(GenomicIntervalVector &amp;grv);
<a name="l00039"></a>00039 };
<a name="l00040"></a>00040 
<a name="l00041"></a>00041 <span class="keyword">typedef</span> vector&lt;AssemblyRegion&gt; AssemblyRegionVector;
<a name="l00042"></a>00042 <span class="keyword">typedef</span> unique_ptr&lt;AssemblyRegion&gt; AssemblyRegionUP;
<a name="l00043"></a>00043 <span class="keyword">typedef</span> vector&lt;AssemblyRegionUP&gt; AssemblyRegionUPVector;
<a name="l00044"></a>00044 
<a name="l00045"></a>00045 
<a name="l00046"></a>00046 <span class="comment">// store a reader of a bam file and vectors of pointers</span>
<a name="l00047"></a>00047 <span class="comment">// to reads which live on the heap, for both anchor and partner</span>
<a name="l00048"></a>00048 <span class="keyword">struct </span>BamAndReads {
<a name="l00049"></a>00049 
<a name="l00050"></a>00050   BamAndReads() {}
<a name="l00051"></a>00051   ~BamAndReads() { 
<a name="l00052"></a>00052 <span class="preprocessor">#ifdef HAVE_BAMTOOLS</span>
<a name="l00053"></a>00053 <span class="preprocessor"></span>    <span class="keyword">delete</span> reader; 
<a name="l00054"></a>00054 <span class="preprocessor">#endif</span>
<a name="l00055"></a>00055 <span class="preprocessor"></span><span class="preprocessor">#ifdef HAVE_HTSLIB</span>
<a name="l00056"></a>00056 <span class="preprocessor"></span>    bgzf_close(fp);                                                                                                                                                                                                 
<a name="l00057"></a>00057     bam_hdr_destroy(br);                                                                                                                                                                                            
<a name="l00058"></a>00058     <span class="comment">//hts_itr_destroy(hts_itr);                                                                                                                                                                                       </span>
<a name="l00059"></a>00059     hts_idx_destroy(idx);                                                                                                                                                                                           
<a name="l00060"></a>00060 <span class="preprocessor">#endif</span>
<a name="l00061"></a>00061 <span class="preprocessor"></span>    arvec.clear();
<a name="l00062"></a>00062 <span class="preprocessor">#ifndef FORGORDON</span>
<a name="l00063"></a>00063 <span class="preprocessor"></span>    <span class="keyword">delete</span> cov;
<a name="l00064"></a>00064 <span class="preprocessor">#endif</span>
<a name="l00065"></a>00065 <span class="preprocessor"></span>    <span class="comment">//for (AssemblyRegionVector::iterator it = arvec.begin(); it != arvec.end(); it++)</span>
<a name="l00066"></a>00066     <span class="comment">//  delete (*it);</span>
<a name="l00067"></a>00067     <span class="comment">//arvec.clear();</span>
<a name="l00068"></a>00068   }
<a name="l00069"></a>00069   BamAndReads(GenomicRegion gr, MiniRulesCollection *tmr, <span class="keywordtype">int</span> tverb, string tbam, string tprefix, <span class="keywordtype">int</span> tlittle, <span class="keywordtype">int</span> tpad);
<a name="l00070"></a>00070 
<a name="l00071"></a>00071 <span class="preprocessor">#ifdef HAVE_BAMTOOLS</span>
<a name="l00072"></a>00072 <span class="preprocessor"></span>  BamReader * reader;
<a name="l00073"></a>00073 <span class="preprocessor">#endif</span>
<a name="l00074"></a>00074 <span class="preprocessor"></span>
<a name="l00075"></a>00075   ReadHash m_all_non_mate_reads, m_all_mate_reads;
<a name="l00076"></a>00076   ReadHash m_allreads; <span class="comment">// vector of pointer to all reads</span>
<a name="l00077"></a>00077 
<a name="l00078"></a>00078   GenomicRegionVector blacklist;
<a name="l00079"></a>00079 
<a name="l00080"></a>00080   <span class="keywordtype">void</span> removeBlacklist(GenomicRegionVector &amp;grv);
<a name="l00081"></a>00081 
<a name="l00082"></a>00082   <span class="keywordtype">int</span> read_time = 0; <span class="comment">// timer in seconds for reading BAM</span>
<a name="l00083"></a>00083   <span class="keywordtype">int</span> unique_reads = 0; <span class="comment">// total number of unique reads to be assembled</span>
<a name="l00084"></a>00084   <span class="keywordtype">int</span> reads = 0; <span class="comment">// total number of reads to be assembled (allows doubles, etc)</span>
<a name="l00085"></a>00085 
<a name="l00086"></a>00086   <span class="keywordtype">int</span> mate_read_time = 0; <span class="comment">// timer in seconds for reading BAM</span>
<a name="l00087"></a>00087   <span class="keywordtype">int</span> mate_unique_reads = 0; <span class="comment">// total number of unique reads to be assembled</span>
<a name="l00088"></a>00088   <span class="keywordtype">int</span> mate_reads = 0; <span class="comment">// total number of reads to be assembled (allows doubles, etc)</span>
<a name="l00089"></a>00089 
<a name="l00090"></a>00090   Coverage * cov = NULL;
<a name="l00091"></a>00091 
<a name="l00092"></a>00092 <span class="preprocessor">#ifdef HAVE_HTSLIB</span>
<a name="l00093"></a>00093 <span class="preprocessor"></span>  <span class="comment">// hts</span>
<a name="l00094"></a>00094   BGZF * fp = 0;
<a name="l00095"></a>00095   hts_idx_t * idx = 0; <span class="comment">// hts_idx_load(bamfile.c_str(), HTS_FMT_BAI);</span>
<a name="l00096"></a>00096   hts_itr_t * hts_itr = 0; <span class="comment">// sam_itr_queryi(idx, 3, 60000, 80000);</span>
<a name="l00097"></a>00097   bam_hdr_t * br = 0; <span class="comment">// header</span>
<a name="l00098"></a>00098   <span class="comment">//samFile* fop = 0;</span>
<a name="l00099"></a>00099   <span class="comment">//fp = sam_open(fn, mode)</span>
<a name="l00100"></a>00100 <span class="preprocessor">#endif  </span>
<a name="l00101"></a>00101 <span class="preprocessor"></span>
<a name="l00102"></a>00102   <span class="comment">// what interval is this defined on</span>
<a name="l00103"></a>00103   GenomicRegion interval;
<a name="l00104"></a>00104   <span class="comment">//GenomicIntervalTreeMap tree;</span>
<a name="l00105"></a>00105   GenomicRegionVector tree;
<a name="l00106"></a>00106 
<a name="l00107"></a>00107   CigarMap cigmap;
<a name="l00108"></a>00108 
<a name="l00109"></a>00109   GenomicRegionVector mate_regions;
<a name="l00110"></a>00110   <span class="comment">//GenomicIntervalTreeMap tree_with_mate;</span>
<a name="l00111"></a>00111 
<a name="l00112"></a>00112   <span class="comment">// define how to read the bam</span>
<a name="l00113"></a>00113   MiniRulesCollection * mr;
<a name="l00114"></a>00114 
<a name="l00115"></a>00115   <span class="keywordtype">int</span> verbose = 1;
<a name="l00116"></a>00116   string bam;
<a name="l00117"></a>00117   string prefix;
<a name="l00118"></a>00118   <span class="keywordtype">int</span> littlechunk = 3000;
<a name="l00119"></a>00119   <span class="keywordtype">int</span> window_pad = 500;
<a name="l00120"></a>00120 
<a name="l00121"></a>00121   <span class="comment">// store all of the mini regions to run</span>
<a name="l00122"></a>00122   AssemblyRegionUPVector arvec;
<a name="l00123"></a>00123 
<a name="l00124"></a>00124   ReadHash getAllReads() { <span class="keywordflow">return</span> m_allreads; }
<a name="l00125"></a>00125 
<a name="l00126"></a>00126   ReadHash getAllNonMateReads() { <span class="keywordflow">return</span> m_all_non_mate_reads; } 
<a name="l00127"></a>00127 
<a name="l00128"></a>00128   <span class="comment">// store all of the discordant reads that map outside of BIGCHUNK</span>
<a name="l00129"></a>00129   <span class="comment">// we'll need these to lookup extra reads for assembly</span>
<a name="l00130"></a>00130   ReadVec disc;
<a name="l00131"></a>00131 
<a name="l00132"></a>00132   <span class="comment">// break up the interval into small pieces</span>
<a name="l00133"></a>00133   <span class="keywordtype">void</span> divideIntoRegions();
<a name="l00134"></a>00134 
<a name="l00135"></a>00135   <span class="comment">// read the bam</span>
<a name="l00136"></a>00136   <span class="keywordtype">void</span> readBam();
<a name="l00137"></a>00137 
<a name="l00138"></a>00138   <span class="comment">// read the bam</span>
<a name="l00139"></a>00139   <span class="keywordtype">void</span> readMateBam();
<a name="l00140"></a>00140 
<a name="l00141"></a>00141   <span class="comment">// get mate regions</span>
<a name="l00142"></a>00142   <span class="keywordtype">void</span> calculateMateRegions();
<a name="l00143"></a>00143 
<a name="l00144"></a>00144   GenomicRegionVector _read_bam(ReadVec &amp;reads, <span class="keywordtype">int</span> limit, <span class="keywordtype">bool</span> cover = <span class="keyword">false</span>);
<a name="l00145"></a>00145 
<a name="l00146"></a>00146   <span class="comment">// use the interval tree to see if a read should be added to a group, based on itself or its mate</span>
<a name="l00147"></a>00147   <span class="keywordtype">void</span> addRead(Read &amp;r);
<a name="l00148"></a>00148   <span class="keywordtype">void</span> addMateRead(Read &amp;r);
<a name="l00149"></a>00149 
<a name="l00150"></a>00150   <span class="keyword">friend</span> ostream&amp; operator&lt;&lt;(ostream &amp;out, BamAndReads &amp;bar);
<a name="l00151"></a>00151 
<a name="l00152"></a>00152 };
<a name="l00153"></a>00153 
<a name="l00154"></a>00154 <span class="preprocessor">#endif</span>
</pre></div><hr size="1"><address style="align: right;"><small>Generated on Thu May 7 11:02:26 2015 for SnowmanSV by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.4.7 </small></address>
</body>
</html>
