<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>SnowmanSV: AlignedContig Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
<link href="tabs.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.4.7 -->
<div class="tabs">
  <ul>
    <li><a href="index.html"><span>Main&nbsp;Page</span></a></li>
    <li id="current"><a href="annotated.html"><span>Classes</span></a></li>
    <li><a href="files.html"><span>Files</span></a></li>
  </ul></div>
<div class="tabs">
  <ul>
    <li><a href="annotated.html"><span>Class&nbsp;List</span></a></li>
    <li><a href="functions.html"><span>Class&nbsp;Members</span></a></li>
  </ul></div>
<h1>AlignedContig Class Reference</h1><!-- doxytag: class="AlignedContig" --><code>#include &lt;<a class="el" href="AlignedContig_8h-source.html">AlignedContig.h</a>&gt;</code>
<p>
<a href="classAlignedContig-members.html">List of all members.</a><table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Public Member Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#bee5720be1588fc0f5e2033c68e47f4b">AlignedContig</a> (const std::string &amp;sam, const BamTools::BamReader *reader, const SnowTools::GenomicRegion &amp;twindow, const CigarMap &amp;nmap, const CigarMap &amp;tmap)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="4b4f577516be62f5f0799f66427cdf84"></a><!-- doxytag: member="AlignedContig::AlignedContig" ref="4b4f577516be62f5f0799f66427cdf84" args="(const BamTools::BamAlignment &amp;a)" -->
&nbsp;</td><td class="memItemRight" valign="bottom"><b>AlignedContig</b> (const BamTools::BamAlignment &amp;a)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#9d4ebb5861cf988ffe1e9aa22715da7f">isWorse</a> (const <a class="el" href="classAlignedContig.html">AlignedContig</a> &amp;ac) const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#adc92b588d491d20316f23d3adebe27c">addAlignment</a> (const BamTools::BamAlignment &amp;align, const SnowTools::GenomicRegion &amp;<a class="el" href="classAlignedContig.html#a94a9f382b760c02333dd2b3085c8c8d">window</a>, const CigarMap &amp;nmap, const CigarMap &amp;tmap)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="bedec8ec48a3179f35d813fb7c013694"></a><!-- doxytag: member="AlignedContig::addAlignment" ref="bedec8ec48a3179f35d813fb7c013694" args="(const BamTools::BamAlignment &amp;align)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><b>addAlignment</b> (const BamTools::BamAlignment &amp;align)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="88a303a9bc8cc07a7efa4548cb73e5fd"></a><!-- doxytag: member="AlignedContig::addDiscordantCluster" ref="88a303a9bc8cc07a7efa4548cb73e5fd" args="(DiscordantCluster dc)" -->
void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#88a303a9bc8cc07a7efa4548cb73e5fd">addDiscordantCluster</a> (DiscordantCluster dc)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">add a discordant cluster that maps to same regions as this contig <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">std::string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#d92b57eafadc8f619a2d9a68a3fee37b">printDiscordantClusters</a> () const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="87074cde30e8ecc61ab848969a7ecc9d"></a><!-- doxytag: member="AlignedContig::getContigName" ref="87074cde30e8ecc61ab848969a7ecc9d" args="() const " -->
std::string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#87074cde30e8ecc61ab848969a7ecc9d">getContigName</a> () const </td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">return the name of the contig <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#6e859c97339b1211923816d20d76380c">getMaxMapq</a> () const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#d339b798afa99c0abcf1a2b5dd27642f">getMinMapq</a> () const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#e5681bd6d2185c4f274d68f83404e072">splitCoverage</a> ()</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#b3889f82451c51a0a8c2946019af6c1a">blacklist</a> (GRC &amp;grv)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#af0765a3cf966158ce94a423682f4e4a">parseDiscovarName</a> (size_t &amp;tumor, size_t &amp;normal)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#de3eff377e9359b6a0a82075530ca638">printContigFasta</a> (std::ofstream &amp;os) const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#d6d2f1dd0a149612a5c9d803a9b6ba85">setMultiMapBreakPairs</a> ()</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#0ad74ad41bc628ebb08ba95c1b6e6376">alignReadsToContigs</a> (ReadVec &amp;bav)</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="54f942ec1c6d4d9f5c5473d145f5fb8a"></a><!-- doxytag: member="AlignedContig::getSequence" ref="54f942ec1c6d4d9f5c5473d145f5fb8a" args="() const " -->
std::string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#54f942ec1c6d4d9f5c5473d145f5fb8a">getSequence</a> () const </td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">return the contig sequence as it came off the assembler <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="f440054db1ee3fabbf7c21a7f6ae5d20"></a><!-- doxytag: member="AlignedContig::hasSubSequence" ref="f440054db1ee3fabbf7c21a7f6ae5d20" args="(const std::string &amp;subseq) const " -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#f440054db1ee3fabbf7c21a7f6ae5d20">hasSubSequence</a> (const std::string &amp;subseq) const </td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">detemine if the contig contains a subsequence <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">bool&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#840897416cdf4e5444b94b3ecfb303ca">hasVariant</a> () const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">std::vector&lt; BreakPoint &gt;&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#0e801da1a6657287362fc291c72a40f3">getAllBreakPoints</a> () const </td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="eb8f9a66cb13b2e2faa56bacf63cf4ad"></a><!-- doxytag: member="AlignedContig::getAllBreakPointPointers" ref="eb8f9a66cb13b2e2faa56bacf63cf4ad" args="() const " -->
std::vector&lt; const BreakPoint * &gt;&nbsp;</td><td class="memItemRight" valign="bottom"><b>getAllBreakPointPointers</b> () const </td></tr>

<tr><td colspan="2"><br><h2>Public Attributes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">std::string&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#a3b7e07742b8d626591bd2078be88920">samrecord</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top">SnowTools::GenomicRegion&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#a94a9f382b760c02333dd2b3085c8c8d">window</a></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="bf3da53e79235032f52481056f2ad1e1"></a><!-- doxytag: member="AlignedContig::m_local_breaks" ref="bf3da53e79235032f52481056f2ad1e1" args="" -->
std::vector&lt; BreakPoint &gt;&nbsp;</td><td class="memItemRight" valign="bottom"><b>m_local_breaks</b></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="aaf8303226b21cd895ff11d55ab0914c"></a><!-- doxytag: member="AlignedContig::m_global_bp" ref="aaf8303226b21cd895ff11d55ab0914c" args="" -->
BreakPoint&nbsp;</td><td class="memItemRight" valign="bottom"><b>m_global_bp</b></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="af838fff09dd3164c41035afd6b18c27"></a><!-- doxytag: member="AlignedContig::m_bamreads" ref="af838fff09dd3164c41035afd6b18c27" args="" -->
ReadVec&nbsp;</td><td class="memItemRight" valign="bottom"><b>m_bamreads</b></td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6485296807f802f3fe2b049d24b47dac"></a><!-- doxytag: member="AlignedContig::m_skip" ref="6485296807f802f3fe2b049d24b47dac" args="" -->
bool&nbsp;</td><td class="memItemRight" valign="bottom"><b>m_skip</b> = false</td></tr>

<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="6372374f9b177011a2c6b43bab9cd4f7"></a><!-- doxytag: member="AlignedContig::m_align" ref="6372374f9b177011a2c6b43bab9cd4f7" args="" -->
AlignmentFragmentVector&nbsp;</td><td class="memItemRight" valign="bottom"><b>m_align</b></td></tr>

<tr><td colspan="2"><br><h2>Friends</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="064b72b01461a676f213a56f80f67d87"></a><!-- doxytag: member="AlignedContig::operator&lt;&lt;" ref="064b72b01461a676f213a56f80f67d87" args="(std::ostream &amp;out, const AlignedContig &amp;ac)" -->
std::ostream &amp;&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="classAlignedContig.html#064b72b01461a676f213a56f80f67d87">operator&lt;&lt;</a> (std::ostream &amp;out, const <a class="el" href="classAlignedContig.html">AlignedContig</a> &amp;ac)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">print this contig <br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
Contains the mapping of an aligned contig to the reference genome, along with pointer to all of the reads aligned to this contig, and a store of all of the breakpoints associated with this contig 
<p>
<hr><h2>Constructor &amp; Destructor Documentation</h2>
<a class="anchor" name="bee5720be1588fc0f5e2033c68e47f4b"></a><!-- doxytag: member="AlignedContig::AlignedContig" ref="bee5720be1588fc0f5e2033c68e47f4b" args="(const std::string &amp;sam, const BamTools::BamReader *reader, const SnowTools::GenomicRegion &amp;twindow, const CigarMap &amp;nmap, const CigarMap &amp;tmap)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">AlignedContig::AlignedContig           </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&nbsp;</td>
          <td class="paramname"> <em>sam</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const BamTools::BamReader *&nbsp;</td>
          <td class="paramname"> <em>reader</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const SnowTools::GenomicRegion &amp;&nbsp;</td>
          <td class="paramname"> <em>twindow</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const CigarMap &amp;&nbsp;</td>
          <td class="paramname"> <em>nmap</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const CigarMap &amp;&nbsp;</td>
          <td class="paramname"> <em>tmap</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Constructor which parses an alignment record from BWA (a potentially multi-line SAM record) <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td>reference to a string representing a SAM alignment (contains newlines if multi-part alignment) </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td>pointer to a BamReader, which is used to convert chr ids to strings (e.g. X, Y) </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td>reference to a SnowTools::GenomicRegion window specifying where in the reference this contig was assembled from. </td></tr>
  </table>
</dl>

</div>
</div><p>
<hr><h2>Member Function Documentation</h2>
<a class="anchor" name="adc92b588d491d20316f23d3adebe27c"></a><!-- doxytag: member="AlignedContig::addAlignment" ref="adc92b588d491d20316f23d3adebe27c" args="(const BamTools::BamAlignment &amp;align, const SnowTools::GenomicRegion &amp;window, const CigarMap &amp;nmap, const CigarMap &amp;tmap)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void AlignedContig::addAlignment           </td>
          <td>(</td>
          <td class="paramtype">const BamTools::BamAlignment &amp;&nbsp;</td>
          <td class="paramname"> <em>align</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const SnowTools::GenomicRegion &amp;&nbsp;</td>
          <td class="paramname"> <em>window</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const CigarMap &amp;&nbsp;</td>
          <td class="paramname"> <em>nmap</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const CigarMap &amp;&nbsp;</td>
          <td class="paramname"> <em>tmap</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Get whether the query is on the reverse strand <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>b</em>&nbsp;</td><td>pointer to an alignment </td></tr>
  </table>
</dl>
<dl compact><dt><b>Returns:</b></dt><dd>boolean true if query is on the reverse strand </dd></dl>

</div>
</div><p>
<a class="anchor" name="0ad74ad41bc628ebb08ba95c1b6e6376"></a><!-- doxytag: member="AlignedContig::alignReadsToContigs" ref="0ad74ad41bc628ebb08ba95c1b6e6376" args="(ReadVec &amp;bav)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void AlignedContig::alignReadsToContigs           </td>
          <td>(</td>
          <td class="paramtype">ReadVec &amp;&nbsp;</td>
          <td class="paramname"> <em>bav</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
align reads to contig and modify their tags to show contig mapping Currently this function will attempt a SmithWaterman alignment for all reads that don't have an exact mapping to the contig. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>bav</em>&nbsp;</td><td>Vector of read smart pointers to align. Modifies their SW tag </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="b3889f82451c51a0a8c2946019af6c1a"></a><!-- doxytag: member="AlignedContig::blacklist" ref="b3889f82451c51a0a8c2946019af6c1a" args="(GRC &amp;grv)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void AlignedContig::blacklist           </td>
          <td>(</td>
          <td class="paramtype">GRC &amp;&nbsp;</td>
          <td class="paramname"> <em>grv</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Checks if any of the indel breaks are in a blacklist. If so, mark the breakpoints of the indels for skipping. That is, hasMinmal() will return false; <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>grv</em>&nbsp;</td><td>The blaclist regions </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="0e801da1a6657287362fc291c72a40f3"></a><!-- doxytag: member="AlignedContig::getAllBreakPoints" ref="0e801da1a6657287362fc291c72a40f3" args="() const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;BreakPoint&gt; AlignedContig::getAllBreakPoints           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const</td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
retrieves all of the breakpoints by combining indels with global mutli-map break <dl compact><dt><b>Returns:</b></dt><dd>vector of ind </dd></dl>

</div>
</div><p>
<a class="anchor" name="6e859c97339b1211923816d20d76380c"></a><!-- doxytag: member="AlignedContig::getMaxMapq" ref="6e859c97339b1211923816d20d76380c" args="() const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int AlignedContig::getMaxMapq           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const<code> [inline]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
get the maximum mapping quality from all alignments <dl compact><dt><b>Returns:</b></dt><dd>int max mapq </dd></dl>

</div>
</div><p>
<a class="anchor" name="d339b798afa99c0abcf1a2b5dd27642f"></a><!-- doxytag: member="AlignedContig::getMinMapq" ref="d339b798afa99c0abcf1a2b5dd27642f" args="() const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int AlignedContig::getMinMapq           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const<code> [inline]</code></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
get the minimum mapping quality from all alignments <dl compact><dt><b>Returns:</b></dt><dd>int min mapq </dd></dl>

</div>
</div><p>
<a class="anchor" name="840897416cdf4e5444b94b3ecfb303ca"></a><!-- doxytag: member="AlignedContig::hasVariant" ref="840897416cdf4e5444b94b3ecfb303ca" args="() const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool AlignedContig::hasVariant           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const</td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
query if this contig contains a potential variant (indel or multi-map) <dl compact><dt><b>Returns:</b></dt><dd>true if there is multimapping or an indel </dd></dl>

</div>
</div><p>
<a class="anchor" name="9d4ebb5861cf988ffe1e9aa22715da7f"></a><!-- doxytag: member="AlignedContig::isWorse" ref="9d4ebb5861cf988ffe1e9aa22715da7f" args="(const AlignedContig &amp;ac) const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool AlignedContig::isWorse           </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classAlignedContig.html">AlignedContig</a> &amp;&nbsp;</td>
          <td class="paramname"> <em>ac</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const</td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
Determine if this contig has identical breaks and is better than another. <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>const</em>&nbsp;</td><td>reference to another <a class="el" href="classAlignedContig.html">AlignedContig</a> </td></tr>
  </table>
</dl>
<dl compact><dt><b>Returns:</b></dt><dd>bool bool returning true iff this contig has identical info has better MAPQ, or equal MAPQ but longer </dd></dl>

</div>
</div><p>
<a class="anchor" name="af0765a3cf966158ce94a423682f4e4a"></a><!-- doxytag: member="AlignedContig::parseDiscovarName" ref="af0765a3cf966158ce94a423682f4e4a" args="(size_t &amp;tumor, size_t &amp;normal)" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool AlignedContig::parseDiscovarName           </td>
          <td>(</td>
          <td class="paramtype">size_t &amp;&nbsp;</td>
          <td class="paramname"> <em>tumor</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t &amp;&nbsp;</td>
          <td class="paramname"> <em>normal</em></td><td>&nbsp;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td><td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
if this is a Discovar contig, extract the tumor and normal read support <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>reference</em>&nbsp;</td><td>to int to fill for normal support </td></tr>
    <tr><td valign="top"></td><td valign="top"><em>reference</em>&nbsp;</td><td>to int to fill for tumor support </td></tr>
  </table>
</dl>
<dl compact><dt><b>Returns:</b></dt><dd>boolean reporting if this is a discovar name </dd></dl>

</div>
</div><p>
<a class="anchor" name="de3eff377e9359b6a0a82075530ca638"></a><!-- doxytag: member="AlignedContig::printContigFasta" ref="de3eff377e9359b6a0a82075530ca638" args="(std::ofstream &amp;os) const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void AlignedContig::printContigFasta           </td>
          <td>(</td>
          <td class="paramtype">std::ofstream &amp;&nbsp;</td>
          <td class="paramname"> <em>os</em>          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const</td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
dump the contigs to a fasta <dl compact><dt><b>Parameters:</b></dt><dd>
  <table border="0" cellspacing="2" cellpadding="0">
    <tr><td valign="top"></td><td valign="top"><em>ostream</em>&nbsp;</td><td>to write to </td></tr>
  </table>
</dl>

</div>
</div><p>
<a class="anchor" name="d92b57eafadc8f619a2d9a68a3fee37b"></a><!-- doxytag: member="AlignedContig::printDiscordantClusters" ref="d92b57eafadc8f619a2d9a68a3fee37b" args="() const " -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string AlignedContig::printDiscordantClusters           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"> const</td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
loop through the vector of DiscordantCluster objects associated with this contig and print 
</div>
</div><p>
<a class="anchor" name="d6d2f1dd0a149612a5c9d803a9b6ba85"></a><!-- doxytag: member="AlignedContig::setMultiMapBreakPairs" ref="d6d2f1dd0a149612a5c9d803a9b6ba85" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void AlignedContig::setMultiMapBreakPairs           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
set the breakpoints on the reference by combining multi-mapped contigs 
</div>
</div><p>
<a class="anchor" name="e5681bd6d2185c4f274d68f83404e072"></a><!-- doxytag: member="AlignedContig::splitCoverage" ref="e5681bd6d2185c4f274d68f83404e072" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void AlignedContig::splitCoverage           </td>
          <td>(</td>
          <td class="paramname">          </td>
          <td>&nbsp;)&nbsp;</td>
          <td width="100%"></td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
loop through all of the breakpoints and calculate the split read support for each. Requires alignedReads to have been run first (will error if not run). 
</div>
</div><p>
<hr><h2>Member Data Documentation</h2>
<a class="anchor" name="a3b7e07742b8d626591bd2078be88920"></a><!-- doxytag: member="AlignedContig::samrecord" ref="a3b7e07742b8d626591bd2078be88920" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string <a class="el" href="classAlignedContig.html#a3b7e07742b8d626591bd2078be88920">AlignedContig::samrecord</a>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
the original SAM record 
</div>
</div><p>
<a class="anchor" name="a94a9f382b760c02333dd2b3085c8c8d"></a><!-- doxytag: member="AlignedContig::window" ref="a94a9f382b760c02333dd2b3085c8c8d" args="" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">SnowTools::GenomicRegion <a class="el" href="classAlignedContig.html#a94a9f382b760c02333dd2b3085c8c8d">AlignedContig::window</a>          </td>
        </tr>
      </table>
</div>
<div class="memdoc">

<p>
reference window from where this contig was assembled 
</div>
</div><p>
<hr>The documentation for this class was generated from the following file:<ul>
<li>/xchip/gistic/Jeremiah/GIT/SnowmanSV/src/Snowman/<a class="el" href="AlignedContig_8h-source.html">AlignedContig.h</a></ul>
<hr size="1"><address style="align: right;"><small>Generated on Thu May 7 11:02:27 2015 for SnowmanSV by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.4.7 </small></address>
</body>
</html>
